README
=head1 NAME

Spreadsheet::Read - Meta-Wrapper for reading spreadsheet data

=head1 SYNOPSYS

  use Spreadsheet::Read;
  my $ref = ReadData ("file.xls");

=head1 DESCRIPTION

Spreadsheet::Read offers a uniformed wrapper to Spreadsheet::ParseExcel
and Spreadheet::ReadSXC to give the end-user a single point of view to
various types of spreadsheets and deal with these in a transparent way.

See for more thorrough documentation the pod in the module, or

  $ man Spreadsheet::Read

after installation

=head1 INSTALLATION

  $ perl Makefile.PL
  $ make
  $ make test
  $ make install

If the C<make test> warns you in the xls tests, read the message and
apply the generated patch. Spreadsheet::ParseExcel has a small bug in
the parsing of the default format regarding UTF-8.

=head1 TODO

  * Make tests for examples/xlscat
  * Field attributes (See Read.pm - TODO)
  * Support Parsers native module options
  * Check if Tk is installed before asking if ss2tk is wanted
  * Test diagnostics output
  * Make clip skip empty sheets

=head1 AUTHOR

H.Merijn Brand, <h.m.brand@xs4all.nl>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2005-2007 H.Merijn Brand

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 

=cut


Changes
Revision history for Spreadsheet::Read

0.29	Tue 21 Oct 2008

    - Make Read safer for files that do not match extension,
      like HTML in foo.xls
    - YAML declared 1.4 (META.yml) instead of 1.1 (YAML)
    - Bring Makefile.PL in sync with META.yml
    - Recommend Text-CSV-0.56

0.28	Wed 04 Sep 2008

    - More tests on date formats
    - Declare Spreadsheet::ParseExcel::FmtDefault in META.yml
    - Add 'size' from Excel font size
    - Update docs about Excel quirks and CPAN links
    - Reference to public git repo
    - Some typo's

0.27	Sun 31 Aug 2008

    - Fixed META.yml (the specs are inconsistent)
    - Added --dtfmt to xlscat

0.26	Fri 29 Aug 2008

    - Added examples/xls2csv
    - Upped copyright to 2008 in examples
    - Don't ask to install examples under automated testing
    - die => croak
    - Added tgzdist target
    - Added encoding options to examples/xlscat
    - Added date-type checks for SS:PE > 0.32

0.25	Wed 02 Jan 2008

    - Spreadsheet-Read now under git
    - Upped copyrights to 2008
    - Added all prereq's to Makefile.PL, even core mods
    - Tested under perl-5.10.0
    - Text::CSV as of 1.00 is OK

0.24	Thu 05 Oct 2007

    - -? and --help in utils moved to Getopt::Long
    - Adjusted copyright notice in utils
    - removed prototypes in utils
    - next is illegal in do {} while (); (Johan Vromans)
      only happens with old Text::CSV_XS

0.23	Thu 21 Jun 2007

    - Use IO::Scalar instead of File::Temp when installed
      Thanks to David Cantrell for making a OpenBSD test box available!
    - use binmode () when opening files
    - Also use 3-arg open in test files
    - die if test files cannot be opene'ed

0.22	Mon 18 Jun 2007

    - 0.21 should already support Text::CSV_PP 1.05
      Tested Text::CSV_PP 1.05 on bleadperl.
    - Better detection of (not) installed modules
    - Module requires perl 5.6.x

0.21	Mon 18 Jun 2007

    - Text::CSV_XS uses keep_meta_info instead of get_flags
    - Removed always_quote from xlscat's CSV output

0.20	Thu 31 May 2007

    - perlcritic OK
    - CSV parsing now uses getline (), and auto-detects eol
      Assumes first line does not contain embedded eol. This implies that
      parsing CSV with embedded eol sequences is now safe.
    - As I now also maintain Text::CSV_XS, I changed the references
    - Changed TODO's from Text::CSV to Text::CSV_PP (::CSV is dated 1997)

0.19	Fri 04 May 2007

    - Could not retreive '0' fields from OpenOffice (Jim Kelly)

0.18	Fri 27 Apr 2007

    - use strict in Makefile.PL
    - Added test_cover make target
    - Added "ods" for OpenOffice conversions
    - Added color tests for ods (but SR::RS doesn't support that yet)
    - Switched from \d to 0-9 in regexp
    - Added more test files and tests (increase coverage)

0.17	Tue 04 Jul 2006

    - xlscat new option: --ansii to (try to) use the ansi colors for fields
    - Fixed a color attribute off-by-one error
    - Added test cases (not yet complete)
    - Added bold and underline

0.16	Tue 04 Jul 2006

    - xlscat new option: --in-sep-char to force-define input sep-char for CSV
    - More debug lines
    - Parser name info in $ref->[0]
    - SquirrelCalc now reports Spreadsheet::Read and its version for parser info
    - Most of the attributes for Excel now implemented. Tested, but no test case

0.15	Wed 21 Jun 2006

    - Small doc change from AnnoCPAN
    - Sheets with undefined labels might cause havoc
    - Clip now skips empty xls sheets (TODO: sxc)
    - xlscat clips by default
    - xlscat new options -d and --noclip
    - xlscat usage () from -?/--help to STDOUT from fault to STDERR

0.14	Fri 20 Jan 2006

    - maxrow and maxcol were swapped in csv sheets
    - promoted internal debug flag to option
    - small doc changes

0.13	Thu 04 Nov 2005

    - Control attrib 'cells' was misinterpreted
    - New option: clip, default is true if {cell} is selected, false otherwise
      Removes trailing lines and columns in each sheet that have no visible data
    - new test t/11_call.t for checking options. Not complete yet
    - Added test_cover target to Makefile.PL

0.12	Mon 31 Oct 2005

    - Added ss2tk to examples

0.11	Wed 26 Oct 2005

    - Allow ods (OpenOffice 2.0) for sxc
    - include real ods test files and new test
    - Check if sc.version is undef, not 0

0.10	Mon 19 Sep 2005

    - More test coverage
    - cr2cell () returns ""     for illegal col/row values
    - cell2cr () returns (0, 0) for illegal cell    values

Makefile.PL
use strict;

use ExtUtils::MakeMaker;

my @exe;
for ( [ "xlscat",	"Convert Spreadsheet to plain text or CSV"	],
      [ "ss2tk",	"Show a Spreadsheet in Perl/Tk"			],
      ) {
    prompt ("Do you want to install '$_->[0]' ($_->[1]) ? ", "n") =~ m/[Yy]/ and
	push @exe, "examples/$_->[0]";
    }
my %wm = (
    NAME         => "Spreadsheet::Read",
    ABSTRACT     => "Read the data from a spreadsheet",
    AUTHOR       => "H.Merijn Brand <h.merijn\@procura.nl>",
    LICENSE	 => "perl",
    VERSION_FROM => "Read.pm",
    EXE_FILES    => [ @exe ],
    PREREQ_FATAL => 0,
    PREREQ_PM    => {
# Backend parsers
	"Text::CSV_XS"			=> 0.00,
	"Spreadsheet::ReadSXC"		=> 0.20,
	"Spreadsheet::ParseExcel"	=> 0.26,
#	"Spreadsheet::Perl"		=> 0.00,	# Not yet

	# Core modules
	"File::Temp"			=> 0.14,

# for ss2tk
#	"Tk"				=> 804.027,
#	"Tk::NoteBook"			=> 4.009,
#	"Tk::TableMatrix::Spreadsheet"	=> 1.2,
	},
    );

my $rv = WriteMakefile (%wm);

package MY;

sub postamble
{
    join "\n" =>
	'cover test_cover:',
	'	cover -delete',
	'	HARNESS_PERL_SWITCHES=-MDevel::Cover $(MAKE) test',
	'	cover',
	'';
    } # postamble

1;


Read.pm
#!/pro/bin/perl

package Spreadsheet::Read;

=head1 NAME

 Spreadsheet::Read - Read the data from a spreadsheet

=head1 SYNOPSYS

 use Spreadsheet::Read;
 my $ref = ReadData ("test.csv", sep => ";");
 my $ref = ReadData ("test.sxc");
 my $ref = ReadData ("test.ods");
 my $ref = ReadData ("test.xls");

 my $a3 = $ref->[1]{A3}, "\n"; # content of field A3 of sheet 1

=cut

use strict;
use warnings;

our $VERSION = "0.24";
sub  Version { $VERSION }

use Exporter;
our @ISA       = qw( Exporter );
our @EXPORT    = qw( ReadData cell2cr cr2cell );
our @EXPORT_OK = qw( parses rows );

use File::Temp   qw( );
use Data::Dumper;

my @parsers = (
    [ csv	=> "Text::CSV_XS"		],
    [ csv	=> "Text::CSV_PP"		],
#   [ csv	=> "Text::CSV"			],	# NYI
    [ ods	=> "Spreadsheet::ReadSXC"	],
    [ sxc	=> "Spreadsheet::ReadSXC"	],
    [ xls	=> "Spreadsheet::ParseExcel"	],
    [ prl	=> "Spreadsheet::Perl"		],

    # Helper modules
    [ ios	=> "IO::Scalar"			],
    );
my %can = map { $_->[0] => 0 } @parsers;
for (@parsers) {
    my ($flag, $mod) = @$_;
    $can{$flag} and next;
    eval "require $mod; \@_ or \$can{\$flag} = '$mod'";
    }
$can{sc} = 1;	# SquirelCalc is built-in

my $debug = 0;
my @def_attr = (
    type    => "text",
    fgcolor => undef,
    bgcolor => undef,
    font    => undef,
    size    => undef,
    format  => undef,
    halign  => "left",
    valign  => "top",
    bold    => 0,
    italic  => 0,
    uline   => 0,
    wrap    => 0,
    merged  => 0,
    hidden  => 0,
    locked  => 0,
    enc     => "utf-8", # $ENV{LC_ALL} // $ENV{LANG} // ...
    );

# Helper functions

# Spreadsheet::Read::parses ("csv") or die "Cannot parse CSV"
sub parses ($)
{
    my $type = shift		or  return 0;
    $type = lc $type;
    # Aliases and fullnames
    $type eq "excel"		and return $can{xls};
    $type eq "oo"		and return $can{sxc};
    $type eq "ods"		and return $can{sxc};
    $type eq "openoffice"	and return $can{sxc};
    $type eq "perl"		and return $can{prl};

    # $can{$type} // 0;
    exists $can{$type} ? $can{$type} : 0;
    } # parses

# cr2cell (4, 18) => "D18"
# No prototype to allow 'cr2cell (@rowcol)'
sub cr2cell
{
    my ($c, $r) = @_;
    defined $c && defined $r && $c > 0 && $r > 0 or return "";
    my $cell = "";
    while ($c) {
	use integer;

	substr ($cell, 0, 0) = chr (--$c % 26 + ord "A");
	$c /= 26;
	}
    "$cell$r";
    } # cr2cell

# cell2cr ("D18") => (4, 18)
sub cell2cr ($)
{
    my ($cc, $r) = ((uc $_[0]) =~ m/^([A-Z]+)([0-9]+)$/) or return (0, 0);
    my $c = 0;
    while ($cc =~ s/^([A-Z])//) {
	$c = 26 * $c + 1 + ord ($1) - ord ("A");
	}
    ($c, $r);
    } # cell2cr

# Convert {cell}'s [column][row] to a [row][column] list
# my @rows = rows ($ss->[1]);
sub rows ($)
{
    my $sheet = shift or return;
    ref    $sheet eq "HASH" && exists $sheet->{cell}   or return;
    exists $sheet->{maxcol} && exists $sheet->{maxrow} or return;
    my $s = $sheet->{cell};

    map {
	my $r = $_;
	[ map { $s->[$_][$r] } 1..$sheet->{maxcol} ];
	} 1..$sheet->{maxrow};
    } # rows

# If option "clip" is set, remove the trailing lines and
# columns in each sheet that contain no visible data
sub _clipsheets
{
    my ($clip, $ref) = @_;
    $clip or return $ref;

    foreach my $sheet (1 .. $ref->[0]{sheets}) {
	my $ss = $ref->[$sheet];

	# Remove trailing empty columns
	while ($ss->{maxcol} and not (
		grep { defined && m/\S/ } @{$ss->{cell}[$ss->{maxcol}]})
		) {
	    (my $col = cr2cell ($ss->{maxcol}, 1)) =~ s/1$//; 
	    my $recol = qr{^$col(?=[0-9]+)$};
	    delete $ss->{$_} for grep m/$recol/, keys %{$ss};
	    $ss->{maxcol}--;
	    }
	$ss->{maxcol} or $ss->{maxrow} = 0;

	# Remove trailing empty lines
	while ($ss->{maxrow} and not (
		grep { defined && m/\S/ }
		map  { $ss->{cell}[$_][$ss->{maxrow}] }
		1 .. $ss->{maxcol}
		)) {
	    my $rerow = qr{^[A-Z]+$ss->{maxrow}$};
	    delete $ss->{$_} for grep m/$rerow/, keys %{$ss};
	    $ss->{maxrow}--;
	    }
	$ss->{maxrow} or $ss->{maxcol} = 0;
	}
    $ref;
    } # _clipsheets

sub _xls_color {
    my ($clr, @clr) = @_;
    @clr == 0 && $clr == 32767 and return undef; # Default fg color
    @clr == 2 && $clr ==     0 and return undef; # No fill bg color
    @clr == 2 && $clr ==     1 and ($clr, @clr) = ($clr[0]);
    @clr and return undef; # Don't know what to do with this
    "#" . lc Spreadsheet::ParseExcel->ColorIdxToRGB ($clr);
    } # _xls_color

sub ReadData ($;@)
{
    my $txt = shift	or  return;
    ref $txt		and return; # TODO: support IO stream (ref $txt eq "IO")

    my $tmpfile;

    my %opt;
    if (@_) {
	   if (ref $_[0] eq "HASH")  { %opt = %{shift @_} }
	elsif (@_ % 2 == 0)          { %opt = @_          }
	}
    defined $opt{rc}	or $opt{rc}	= 1;
    defined $opt{cells}	or $opt{cells}	= 1;
    defined $opt{attr}	or $opt{attr}	= 0;
    defined $opt{clip}	or $opt{clip}	= $opt{cell};

    # $debug = $opt{debug} // 0;
    $debug = defined $opt{debug} ? $opt{debug} : 0;
    $debug > 4 and print STDERR Data::Dumper->Dump ([\%opt],["Options"]);

    # CSV not supported from streams
    if ($txt =~ m/\.(csv)$/i and -f $txt) {
	$can{csv} or die "CSV parser not installed";

	$debug and print STDERR "Opening CSV $txt\n";
	open my $in, "<", $txt or return;
	my $csv;
	my @data = (
	    {	type	=> "csv",
		parser  => $can{csv},
		version	=> $can{csv}->VERSION,
		quote   => '"',
		sepchar => ',',
		sheets	=> 1,
		sheet	=> { $txt => 1 },
		},
	    {	label	=> $txt,
		maxrow	=> 0,
		maxcol	=> 0,
		cell	=> [],
		attr	=> [],
		},
	    );
	$_ = <$in>;
	my $quo = defined $opt{quote} ? $opt{quote} : '"';
	my $sep = # If explicitely set, use it
	   defined $opt{sep} ? $opt{sep} :
	       # otherwise start auto-detect with quoted strings
	       m/["0-9];["0-9;]/	? ";"  :
	       m/["0-9],["0-9,]/	? ","  :
	       m/["0-9]\t["0-9,]/	? "\t" :
	       # If neither, then for unquoted strings
	       m/\w;[\w;]/		? ";"  :
	       m/\w,[\w,]/		? ","  :
	       m/\w\t[\w,]/		? "\t" :
					  ","  ;
	my ($eol) = m{([\r\n]+)\z};
	$debug > 1 and print STDERR "CSV sep_char '$sep', quote_char '$quo'\n";
	$csv = $can{csv}->new ({
	    sep_char       => ($data[0]{sepchar} = $sep),
	    quote_char     => ($data[0]{quote}   = $quo),
	    eol            => $eol,
	    keep_meta_info => 1,	# Ignored for Text::CSV_XS <= 0.27
	    binary         => 1,
	    }) or die "Cannot create a csv ('$sep', '$quo', '$eol') parser!";

	# while ($row = $csv->getline () {
	# doesn't work, because I have to fetch the first line for auto
	# detecting sep and quote
	$csv->parse ($_);
	my $row = [ $csv->fields ];
	do {
	    if (my @row = @$row) {
		my $r = ++$data[1]{maxrow};
		@row > $data[1]{maxcol} and $data[1]{maxcol} = @row;
		foreach my $c (0 .. $#row) {
		    my $val = $row[$c];
		    my $cell = cr2cell ($c + 1, $r);
		    $opt{rc}    and $data[1]{cell}[$c + 1][$r] = $val;
		    $opt{cells} and $data[1]{$cell} = $val;
		    $opt{attr}  and $data[1]{attr}[$c + 1][$r] = { @def_attr };
		    }
		}
	    else {
		$csv = undef;
		}
	    } while ($csv && ($row = $csv->getline ($in)));
	close $in;

	for (@{$data[1]{cell}}) {
	    defined $_ or $_ = [];
	    }
	return _clipsheets $opt{clip}, [ @data ];
	}

    # From /etc/magic: Microsoft Office Document
    my $xls_from_txt;
    if ($txt =~ m/^(\376\067\0\043
		   |\320\317\021\340\241\261\032\341
		   |\333\245-\0\0\0)/x) {
	$can{xls} or die "Spreadsheet::ParseExcel not installed";
	if ($can{ios}) { # Do not use a temp file if IO::Scalar is available
	    $xls_from_txt = \$txt;
	    }
	else {
	    $tmpfile = File::Temp->new (SUFFIX => ".xls", UNLINK => 1);
	    binmode $tmpfile;
	    print   $tmpfile $txt;
	    $txt = "$tmpfile";
	    }
	}
    if ($xls_from_txt or $txt =~ m/\.xls$/i && -f $txt) {
	$can{xls} or die "Spreadsheet::ParseExcel not installed";
	my $oBook;
	if ($xls_from_txt) {
	    $debug and print STDERR "Opening XLS \$txt\n";
	    $oBook = Spreadsheet::ParseExcel::Workbook->Parse (\$txt);
	    }
	else {
	    $debug and print STDERR "Opening XLS $txt\n";
	    $oBook = Spreadsheet::ParseExcel::Workbook->Parse ($txt);
	    }
	$oBook or return;
	$debug > 8 and print STDERR Data::Dumper->Dump ([$oBook],["oBook"]);
	my @data = ( {
	    type	=> "xls",
	    parser	=> "Spreadsheet::ParseExcel",
	    version	=> $Spreadsheet::ParseExcel::VERSION,
	    sheets	=> $oBook->{SheetCount},
	    sheet	=> {},
	    } );
	$debug and print STDERR "\t$data[0]{sheets} sheets\n";
	foreach my $oWkS (@{$oBook->{Worksheet}}) {
	    $opt{clip} and !defined $oWkS->{Cells} and next; # Skip empty sheets
	    my %sheet = (
		label	=> $oWkS->{Name},
		maxrow	=> 0,
		maxcol	=> 0,
		cell	=> [],
		attr	=> [],
		);
	    defined $sheet{label}  or  $sheet{label}  = "-- unlabeled --";
	    exists $oWkS->{MaxRow} and $sheet{maxrow} = $oWkS->{MaxRow} + 1;
	    exists $oWkS->{MaxCol} and $sheet{maxcol} = $oWkS->{MaxCol} + 1;
	    my $sheet_idx = 1 + @data;
	    $debug and print STDERR "\tSheet $sheet_idx '$sheet{label}' $sheet{maxrow} x $sheet{maxcol}\n";
	    if (exists $oWkS->{MinRow}) {
		foreach my $r ($oWkS->{MinRow} .. $sheet{maxrow}) { 
		    foreach my $c ($oWkS->{MinCol} .. $sheet{maxcol}) { 
			my $oWkC = $oWkS->{Cells}[$r][$c] or next;
			defined (my $val = $oWkC->{Val})  or next;
			my $cell = cr2cell ($c + 1, $r + 1);
			$opt{rc}    and $sheet{cell}[$c + 1][$r + 1] = $val;	# Original
			$opt{cells} and $sheet{$cell} = $oWkC->Value;	# Formatted
			if ($opt{attr}) {
			    my $FmT = $oWkC->{Format};
			    my $FnT = $FmT->{Font};
			    $sheet{attr}[$c + 1][$r + 1] = {
				@def_attr,

				type    => lc $oWkC->{Type},
				enc     => $oWkC->{Code},
				merged	=> $oWkC->{Merged} || 0,
				hidden	=> $FmT->{Hidden},
				locked	=> $FmT->{Lock},
				format  => $FmT->{FmtIdx}
					   ? $oBook->{FormatStr}{$FmT->{FmtIdx}}
					   : undef,
				halign  => [ undef, qw( left center right
					    fill justify ), undef,
					    "equal_space" ]->[$FmT->{AlignH}],
				valign  => [ qw( top center bottom justify
					    equal_space )]->[$FmT->{AlignV}],
				wrap    => $FmT->{Wrap},
				font    => $FnT->{Name},
				bold	=> $FnT->{Bold},
				italic	=> $FnT->{Italic},
				uline	=> $FnT->{Underline},
				fgcolor => _xls_color ($FnT->{Color}),
				bgcolor => _xls_color (@{$FmT->{Fill}}),
				};
			    }
			}
		    }
		}
	    for (@{$sheet{cell}}) {
		defined $_ or $_ = [];
		}
	    push @data, { %sheet };
#	    $data[0]{sheets}++;
	    if ($sheet{label} eq "-- unlabeled --") {
		$sheet{label} = "";
		}
	    else {
		$data[0]{sheet}{$sheet{label}} = $#data;
		}
	    }
	return _clipsheets $opt{clip}, [ @data ];
	}

    if ($txt =~ m/^# .*SquirrelCalc/ or $txt =~ m/\.sc$/ && -f $txt) {
	if ($txt !~ m/\n/ && -f $txt) {
	    local $/;
	    open my $sc, "<", $txt or return;
	    $txt = <$sc>;
	    $txt =~ m/\S/ or return;
	    }
	my @data = (
	    {	type	=> "sc",
		parser	=> "Spreadsheet::Read",
		version	=> $VERSION,
		sheets	=> 1,
		sheet	=> { sheet => 1 },
		},
	    {	label	=> "sheet",
		maxrow	=> 0,
		maxcol	=> 0,
		cell	=> [],
		attr	=> [],
		},
	    );

	for (split m/\s*[\r\n]\s*/, $txt) {
	    if (m/^dimension.*of ([0-9]+) rows.*of ([0-9]+) columns/i) {
		@{$data[1]}{qw(maxrow maxcol)} = ($1, $2);
		next;
		}
	    s/^r([0-9]+)c([0-9]+)\s*=\s*// or next;
	    my ($c, $r) = map { $_ + 1 } $2, $1;
	    if (m/.* {(.*)}$/ or m/"(.*)"/) {
		my $cell = cr2cell ($c, $r);
		$opt{rc}    and $data[1]{cell}[$c][$r] = $1;
		$opt{cells} and $data[1]{$cell} = $1;
		$opt{attr}  and $data[1]{attr}[$c + 1][$r] = { @def_attr };
		next;
		}
	    # Now only formula's remain. Ignore for now
	    # r67c7 = [P2L] 2*(1000*r67c5-60)
	    }
	for (@{$data[1]{cell}}) {
	    defined $_ or $_ = [];
	    }
	return _clipsheets $opt{clip}, [ @data ];
	}

    if ($txt =~ m/^<\?xml/ or -f $txt) {
	$can{sxc} or die "Spreadsheet::ReadSXC not installed";
	my $sxc_options = { OrderBySheet => 1 }; # New interface 0.20 and up
	my $sxc;
	   if ($txt =~ m/\.(sxc|ods)$/i) {
	    $debug and print STDERR "Opening \U$1\E $txt\n";
	    $sxc = Spreadsheet::ReadSXC::read_sxc      ($txt, $sxc_options)	or  return;
	    }
	elsif ($txt =~ m/\.xml$/i) {
	    $debug and print STDERR "Opening XML $txt\n";
	    $sxc = Spreadsheet::ReadSXC::read_xml_file ($txt, $sxc_options)	or  return;
	    }
	# need to test on pattern to prevent stat warning
	# on filename with newline
	elsif ($txt !~ m/^<\?xml/i and -f $txt) {
	    $debug and print STDERR "Opening XML $txt\n";
	    open my $f, "<", $txt	or  return;
	    local $/;
	    $txt = <$f>;
	    }
	!$sxc && $txt =~ m/^<\?xml/i and
	    $sxc = Spreadsheet::ReadSXC::read_xml_string ($txt, $sxc_options);
	$debug > 8 and print STDERR Data::Dumper->Dump ([$sxc],["sxc"]);
	if ($sxc) {
	    my @data = ( {
		type	=> "sxc",
		parser	=> "Spreadsheet::ReadSXC",
		version	=> $Spreadsheet::ReadSXC::VERSION,
		sheets	=> 0,
		sheet	=> {},
		} );
	    my @sheets = ref $sxc eq "HASH"	# < 0.20
		? map {
		    {   label => $_,
			data  => $sxc->{$_},
			}
		    } keys %$sxc
		: @{$sxc};
	    foreach my $sheet (@sheets) {
		my @sheet = @{$sheet->{data}};
		my %sheet = (
		    label	=> $sheet->{label},
		    maxrow	=> scalar @sheet,
		    maxcol	=> 0,
		    cell	=> [],
		    attr	=> [],
		    );
		my $sheet_idx = 1 + @data;
		$debug and print STDERR "\tSheet $sheet_idx '$sheet{label}' $sheet{maxrow} rows\n";
		foreach my $r (0 .. $#sheet) {
		    my @row = @{$sheet[$r]} or next;
		    foreach my $c (0 .. $#row) {
			defined (my $val = $row[$c]) or next;
			my $C = $c + 1;
			$C > $sheet{maxcol} and $sheet{maxcol} = $C;
			my $cell = cr2cell ($C, $r + 1);
			$opt{rc}    and $sheet{cell}[$C][$r + 1] = $val;
			$opt{cells} and $sheet{$cell} = $val;
			$opt{attr}  and $sheet{attr}[$C][$r + 1] = { @def_attr };
			}
		    }
		for (@{$sheet{cell}}) {
		    defined $_ or $_ = [];
		    }
		$debug and print STDERR "\tSheet $sheet_idx '$sheet{label}' $sheet{maxrow} x $sheet{maxcol}\n";
		push @data, { %sheet };
		$data[0]{sheets}++;
		$data[0]{sheet}{$sheet->{label}} = $#data;
		}
	    return _clipsheets $opt{clip}, [ @data ];
	    }
	}

    return;
    } # ReadData

1;

=head1 DESCRIPTION

Spreadsheet::Read tries to transparently read *any* spreadsheet and
return its content in a universal manner independent of the parsing
module that does the actual spreadsheet scanning.

For OpenOffice this module uses Spreadsheet::ReadSXC

For Excel this module uses Spreadsheet::ParseExcel

For CSV this module uses Text::CSV_XS (0.29 or up prefered) or
Text_PP (1.05 or up required).

For SquirrelCalc there is a very simplistic built-in parser

=head2 Data structure

The data is returned as an array reference:

  $ref = [
 	# Entry 0 is the overall control hash
 	{ sheets  => 2,
	  sheet   => {
	    "Sheet 1"	=> 1,
	    "Sheet 2"	=> 2,
	    },
	  type    => "xls",
	  parser  => "Spreadsheet::ParseExcel",
	  version => 0.26,
	  },
 	# Entry 1 is the first sheet
 	{ label  => "Sheet 1",
 	  maxrow => 2,
 	  maxcol => 4,
 	  cell   => [ undef,
	    [ undef, 1 ],
	    [ undef, undef, undef, undef, undef, "Nugget" ],
	    ],
 	  A1     => 1,
 	  B5     => "Nugget",
 	  },
 	# Entry 2 is the second sheet
 	{ label => "Sheet 2",
 	  :
 	:

To keep as close contact to spreadsheet users, row and column 1 have
index 1 too in the C<cell> element of the sheet hash, so cell "A1" is
the same as C<cell> [1, 1] (column first). To switch between the two,
there are two helper functions available: C<cell2cr ()> and C<cr2cell ()>.

The C<cell> hash entry contains unformatted data, while the hash entries
with the traditional labels contain the formatted values (if applicable).

The control hash (the first entry in the returned array ref), contains
some spreadsheet meta-data. The entry C<sheet> is there to be able to find
the sheets when accessing them by name:

  my %sheet2 = %{$ref->[$ref->[0]{sheet}{"Sheet 2"}]};

=head2 Functions

=over 2

=item my $ref = ReadData ($source [, option => value [, ... ]]);

=item my $ref = ReadData ("file.csv", sep => ',', quote => '"');

=item my $ref = ReadData ("file.xls");

=item my $ref = ReadData ("file.ods");

=item my $ref = ReadData ("file.sxc");

=item my $ref = ReadData ("content.xml");

=item my $ref = ReadData ($content);

Tries to convert the given file, string, or stream to the data
structure described above.

Precessing data from a stream or content is supported for Excel
(through a File::Temp temporary file or IO::Scalar when available),
or for XML (OpenOffice), but not for CSV.

ReadSXC does preserve sheet order as of version 0.20.

Currently supported options are:

=over 2

=item cells

Control the generation of named cells ("A1" etc). Default is true.

=item rc

Control the generation of the {cell}[c][r] entries. Default is true.

=item attr

Control the generation of the {attr}[c][r] entries. Default is false.

=item clip

If set, C<ReadData ()> will remove all trailing lines and columns per
sheet that have no visual data.
This option is only valid if C<cells> is true. The default value is
true if C<cells> is true, and false otherwise.

=item sep

Set separator for CSV. Default is comma C<,>.

=item quote

Set quote character for CSV. Default is C<">.

=item debug

Enable some diagnostic messages to STDERR.

The value determines how much diagnostics are dumped (using Data::Dumper).
A value of 9 and higher will dump the entire structure from the backend
parser.

=back

In case of CSV parsing, C<ReadData ()> will use the first line to
auto-detect the separation character, if not explicitly passed, and the
end-of-line sequence. This means that if the first line does not contain
embedded newlines, the rest of the CSV file can have them, and they will
be parsed correctly.

=item my $cell = cr2cell (col, row)

C<cr2cell ()> converts a C<(column, row)> pair (1 based) to the
traditional cell notation:

  my $cell = cr2cell ( 4, 14); # $cell now "D14"
  my $cell = cr2cell (28,  4); # $cell now "AB4"

=item my ($col, $row) = cell2cr ($cell)

C<cell2cr ()> converts traditional cell notation to a C<(column, row)>
pair (1 based):

  my ($col, $row) = cell2cr ("D14"); # returns ( 4, 14)
  my ($col, $row) = cell2cr ("AB4"); # returns (28,  4)

=item my @rows = rows ($ref)

=item my @rows = Spreadsheet::Read::rows ($ss->[1])

Convert C<{cell}>'s C<[column][row]> to a C<[row][column]> list.

Note that the indexes in the returned list are 0-based, where the
index in the C<{cell}> entry is 1-based.

C<rows ()> is not imported by default, so either specify it in the
use argument list, or call it fully qualified.

=item parses ($format)

=item Spreadsheet::Read::parses ("CSV")

C<parses ()> returns Spreadsheet::Read's capability to parse the
required format.

C<parses ()> is not imported by default, so either specify it in the
use argument list, or call it fully qualified.

=item my $rs_version = Version ()

=item my $v = Spreadsheet::Read::Version ()

Returns the current version of Spreadsheet::Read.

C<Version ()> is not imported by default, so either specify it in the
use argument list, or call it fully qualified.

=back

=head1 TODO

=over 4

=item Cell attributes

Future plans include cell attributes, available as for example:

        { label  => "Sheet 1",
          maxrow => 2,
          maxcol => 4,
          cell   => [ undef,
            [ undef, 1 ],
            [ undef, undef, undef, undef, undef, "Nugget" ],
            ],
          attr   => [ undef,
            [ undef, {
              type    => "numeric",
              fgcolor => "#ff0000",
              bgcolor => undef,
              font    => "Arial",
              size    => undef,
              format  => "## ##0.00",
              halign  => "right",
              valign  => "top",
              uline   => 0,
              bold    => 0,
              italic  => 0,
              wrap    => 0,
              merged  => 0,
              hidden  => 0,
              locked  => 0,
              enc     => "utf-8",
              }, ]
            [ undef, undef, undef, undef, undef, {
              type    => "text",
 	      fgcolor => "#e2e2e2",
              bgcolor => undef,
              font    => "Letter Gothic",
              size    => 15,
              format  => undef,
              halign  => "left",
              valign  => "top",
              uline   => 0,
              bold    => 0,
              italic  => 0,
              wrap    => 0,
              merged  => 0,
              hidden  => 0,
              locked  => 0,
 	      enc     => "iso8859-1",
 	      }, ]
 	  A1     => 1,
 	  B5     => "Nugget",
 	  },

This has now been partially implemented. Excel only.

=item Options

=over 2

=item Module Options

New Spreadsheet::Read options are bound to happen. I'm thinking of an
option that disables the reading of the data entirely to speed up an
index request (how many sheets/fields/columns). See C<xlscat -i>.

=item Parser options

Try to transparently support as many options as the encapsulated modules
support regarding (un)formatted values, (date) formats, hidden columns
rows or fields etc. These could be implemented like C<attr> above but
names C<meta>, or just be new values in the C<attr> hashes.

=back

=item Other spreadsheet formats

I consider adding any spreadsheet interface that offers a usable API.

=item OO-ify

Consider making the ref an object, though I currently don't see the big
advantage (yet). Maybe I'll make it so that it is a hybrid functional /
OO interface.

=back

=head1 SEE ALSO

=over 2

=item Text::CSV_XS, Text::CSV_PP

http://search.cpan.org/~hmbrand/

A pure perl version is available on http://search.cpan.org/~makamaka/

=item Spreadsheet::ParseExcel

http://search.cpan.org/~kwitknr/

=item Spreadsheet::ReadSXC

http://search.cpan.org/~terhechte/

=item Spreadsheet::BasicRead

http://search.cpan.org/~gng/ for xlscat likewise functionality (Excel only)

=item Spreadsheet::ConvertAA

http://search.cpan.org/~nkh/ for an alternative set of cell2cr () /
cr2cell () pair

=item Spreadsheet::Perl

http://search.cpan.org/~nkh/ offers a Pure Perl implementation of a
spreadsheet engine. Users that want this format to be supported in
Spreadsheet::Read are hereby motivated to offer patches. It's not high
on my todo-list.

=item xls2csv

http://search.cpan.org/~ken/ offers an alternative for my C<xlscat -c>,
in the xls2csv tool, but this tool focusses on character encoding
transparency, and requires some other modules.

=back

=head1 AUTHOR

H.Merijn Brand, <h.m.brand@xs4all.nl>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2005-2007 H.Merijn Brand

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 

=cut
